spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkValueConverter.java

```diff
@@ -72,9 +73,16 @@ public static Object convert(Type type, Object object) {
         // for Spark SQL DATE type otherwise java.sql.Date is returned.
         return DateTimeUtils.anyToDays(object);
       case TIMESTAMP:
-        // if spark.sql.datetime.java8API.enabled is set to true, java.time.Instant
-        // for Spark SQL TIMESTAMP type is returned otherwise java.sql.Timestamp is returned.
-        return DateTimeUtils.anyToMicros(object);
+        Types.TimestampType ts = (Types.TimestampType) type.asPrimitiveType();
+
+        if (ts.shouldAdjustToUTC()) {
+          // if spark.sql.datetime.java8API.enabled is set to true, java.time.Instant
+          // for Spark SQL TIMESTAMP type is returned otherwise java.sql.Timestamp is returned.
+          return DateTimeUtils.anyToMicros(object);
+        } else {
+          LocalDateTime ldt = (LocalDateTime) object;
+          return DateTimeUtils.localDateTimeToMicros(ldt);
```

Comment:
Why not just add `LocalDateTime` support to `anyToMicros`? That seems to make more sense than assuming a `LocalDateTime` for `timestampntz`.

Reply:
I like it a lot: https://github.com/apache/spark/pull/41238